contracts/src/batch-bridge/L1BatchBridgeGateway.sol

```diff
@@ -0,0 +1,418 @@
+        }
+        tokens[token] = cachedTokenState;
+
+        // refund keeper fee
+        unchecked {
+            if (msg.value > depositFee + batchBridgeFee) {
+                _transferToken(address(0), _msgSender(), msg.value - depositFee - batchBridgeFee);
+            }
+        }
+    }
+
+    /**********************
+     * Internal Functions *
+     **********************/
+
+    /// @dev Internal function to deposit token.
+    /// @param token The address of token to deposit.
+    /// @param sender The address of token sender.
+    /// @param amount The amount of token to deposit.
+    function _deposit(
+        address token,
+        address sender,
+        uint96 amount
+    ) internal nonReentrant {
+        BatchSetting memory cachedBatchSetting = settings[token];
+        TokenState memory cachedTokenState = tokens[token];
+        _tryFinalzeCurrentPhase(token, cachedBatchSetting, cachedTokenState);
+
+        PhaseState memory cachedPhaseState = phases[token][cachedTokenState.currentPhaseIndex];
+        if (amount < cachedBatchSetting.minAmountPerTx) {
+            revert ErrorDepositAmountTooSmall();
+        }
+
+        // deduct fee and update cached state
+        unchecked {
+            amount -= cachedBatchSetting.feeAmountPerTx;
+            cachedTokenState.pending += amount;
+            cachedPhaseState.amount += amount;
+            cachedPhaseState.numDeposits += 1;
+        }
+
+        // compute the hash chain
+        bytes32 node = BatchBridgeCodec.encodeNode(sender, amount);
+        if (cachedPhaseState.hash == bytes32(0)) {
+            bytes32 initialNode = BatchBridgeCodec.encodeInitialNode(token, cachedTokenState.currentPhaseIndex);
+            // this is first tx in this phase
+            cachedPhaseState.hash = BatchBridgeCodec.hash(initialNode, node);
+            cachedPhaseState.firstDepositTimestamp = uint64(block.timestamp);
+        } else {
+            cachedPhaseState.hash = BatchBridgeCodec.hash(cachedPhaseState.hash, node);
+        }
+
+        phases[token][cachedTokenState.currentPhaseIndex] = cachedPhaseState;
+        tokens[token] = cachedTokenState;
+    }
+
+    /// @dev Internal function to finalze current phase.
+    ///      This function may change the value of `cachedTokenState`, which can be used in later operation.
+    /// @param token The address of token to finalze.
+    /// @param cachedBatchSetting The cached batch setting in memory.
+    /// @param cachedTokenState The cached token state in memory.
+    function _tryFinalzeCurrentPhase(
+        address token,
+        BatchSetting memory cachedBatchSetting,
+        TokenState memory cachedTokenState
+    ) internal view {
+        if (cachedBatchSetting.maxTxPerBatch == 0) {
+            revert ErrorTokenNotSupported();
+        }
+        PhaseState memory cachedPhaseState = phases[token][cachedTokenState.currentPhaseIndex];
+        if (cachedPhaseState.numDeposits == 0) return;
```

Comment:
this line seems unnecessary

Reply:
This line is used to skip the very first batch. In the first batch, `firstDepositTimestamp` is zero. If we won't add this line, the first batch will be considered finalized.